pkg/reconciler/taskrun/metrics_test.go

```diff
@@ -0,0 +1,288 @@
+package taskrun
+
+import (
+	"testing"
+	"time"
+
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+	alpha1 "github.com/tektoncd/pipeline/pkg/client/informers/externalversions/pipeline/v1alpha1"
+	faketaskruninformer "github.com/tektoncd/pipeline/pkg/client/injection/informers/pipeline/v1alpha1/taskrun/fake"
+	tb "github.com/tektoncd/pipeline/test/builder"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/metrics/metricstest"
+	rtesting "knative.dev/pkg/reconciler/testing"
+)
+
+func TestUninitializedMetrics(t *testing.T) {
+	metrics := Recorder{}
+
+	durationCountError := metrics.DurationAndCount(&v1alpha1.TaskRun{})
+	taskrunsCountError := metrics.RunningTaskRuns(nil)
+	podLatencyError := metrics.RecordPodLatency(nil, nil)
+
+	assertErrNotNil(durationCountError, "DurationCount recording expected to return error but got nil", t)
+	assertErrNotNil(taskrunsCountError, "Current TaskrunsCount recording expected to return error but got nil", t)
+	assertErrNotNil(podLatencyError, "Pod Latency recording expected to return error but got nil", t)
+}
+
+func TestRecordTaskrunDurationCount(t *testing.T) {
+	startTime := time.Now()
+
+	testData := []struct {
+		name             string
+		taskRun          *v1alpha1.TaskRun
+		expectedTags     map[string]string
+		expectedDuration float64
+		expectedCount    int64
+	}{{
+		name: "for_succeeded_task",
+		taskRun: tb.TaskRun("taskrun-1", "ns",
+			tb.TaskRunSpec(
+				tb.TaskRunTaskRef("task-1"),
+			),
+			tb.TaskRunStatus(
+				tb.TaskRunStartTime(startTime),
+				tb.TaskRunCompletionTime(startTime.Add(1*time.Minute)),
+				tb.StatusCondition(apis.Condition{
+					Type:   apis.ConditionSucceeded,
+					Status: corev1.ConditionTrue,
+				}),
+			)),
+		expectedTags: map[string]string{
+			"task":      "task-1",
+			"taskrun":   "taskrun-1",
+			"namespace": "ns",
+			"status":    "success",
+		},
+		expectedDuration: 60,
+		expectedCount:    1,
+	}, {
+		name: "for_failed_task",
+		taskRun: tb.TaskRun("taskrun-1", "ns",
+			tb.TaskRunSpec(
+				tb.TaskRunTaskRef("task-1"),
+			),
+			tb.TaskRunStatus(
+				tb.TaskRunStartTime(startTime),
+				tb.TaskRunCompletionTime(startTime.Add(1*time.Minute)),
+				tb.StatusCondition(apis.Condition{
+					Type:   apis.ConditionSucceeded,
+					Status: corev1.ConditionFalse,
+				}),
+			)),
+		expectedTags: map[string]string{
+			"task":      "task-1",
+			"taskrun":   "taskrun-1",
+			"namespace": "ns",
+			"status":    "failed",
+		},
+		expectedDuration: 60,
+		expectedCount:    1,
+	}}
+
+	for _, test := range testData {
+		t.Run(test.name, func(t *testing.T) {
+			defer unregisterMetrics()
+
+			metrics, _ := NewRecorder()
```

Comment:
Here too (and below)

Reply:
ahh my bad, thank you for being patient :bowing_man: